src/Stratis.Bitcoin/Utilities/AsyncExecutionEvent.cs

```diff
@@ -0,0 +1,202 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    /// <summary>
+    /// Asynchronous event handler that can be registered with <see cref="AsyncExecutionEvent{TSender, TArg}"/>.
+    /// </summary>
+    /// <typeparam name="TSender">Type of the sender object that is the source of the event.</typeparam>
+    /// <typeparam name="TArg">Type of the argument that is passed to the callback.</typeparam>
+    /// <param name="sender">Source of the event.</param>
+    /// <param name="arg">Callback argument.</param>
+    public delegate Task AsyncExecutionEventCallback<TSender, TArg>(TSender sender, TArg arg);
+
+    /// <summary>
+    /// Execution event is a specific moment in the execution flow of that a component 
+    /// that other components are allowed to be subscribed to and get notified about 
+    /// when it occurs.
+    /// <para>
+    /// This implementation allows components to register asynchronous event handlers.
+    /// </para>
+    /// </summary>
+    /// <typeparam name="TSender">Type of event source sender objects.</typeparam>
+    /// <typeparam name="TArg">Type of arguments that are passed to callbacks.</typeparam>
+    public class AsyncExecutionEvent<TSender, TArg>: IDisposable
+    {
+        /// <summary>
+        /// Protects access to <see cref="callbackList"/> and <see cref="callbackToListNodeMapping"/>,
+        /// and also provides guarantees of <see cref="Unregister(AsyncExecutionEventCallback{TSender, TArg})"/> method.
+        /// </summary>
+        private readonly AsyncLock lockObject;
+
+        /// <summary>List of registered callbacks.</summary>
+        /// <remarks>All access to this object has to be protected with <see cref="lockObject"/>.</remarks>
+        private readonly LinkedList<AsyncExecutionEventCallback<TSender, TArg>> callbackList;
+
+        /// <summary>Mapping of registered callbacks to nodes of <see cref="callbackList"/> to allow fast lookup during removals.</summary>
+        /// <remarks>All access to this object has to be protected with <see cref="lockObject"/>.</remarks>
+        private readonly Dictionary<AsyncExecutionEventCallback<TSender, TArg>, LinkedListNode<AsyncExecutionEventCallback<TSender, TArg>>> callbackToListNodeMapping;
+
+        /// <summary>
+        /// Set to <c>true</c> if the current async execution context is the one that executes the callbacks,
+        /// set to <c>false</c> otherwise.
+        /// </summary>
+        /// <remarks>
+        /// This allows <see cref="Register(AsyncExecutionEventCallback{TSender, TArg}, bool)"/> and <see cref="Unregister(AsyncExecutionEventCallback{TSender, TArg})"/>
+        /// to recognize whether they are executing from a callback or not.
+        /// </remarks>
```

Comment:
it will unregister, this is a major fuckup in whole code, behaviors do call dispose/unregister stuff from within callbacks, so you end up with a ton of cyclic calls like

state changed -> on state changed callback -> dispose peer -> disconnect peer ->state changed to offline -> on state changed callback -> dispose peer

Reply:
i even had to change async queue (see commit below) to allow this because of deadlock again

the requirements the current code has on these primitives is quite crazy
it seems quite lazy to me to allow such things as disposing/unregistering when you are inside the callback, the user code is simpler with that, but to argue about the code is pretty hard; but NP refactoring can't fix this, this is in all code, in all kinds of different behaviors, we just do that - when something changes, we want the peer to be disconnected and we do not have it separated from disposal. so to really fix this in a nice way, we would need to separate disposals from disconnects in all code and all code should properly manage their peer objects ...

Reply:
not saying it is impossible, it just does not belong to NP refactoring task, and so for me at this point, the only solution was to allow such things

Reply:
created issue for that #1090

Reply:
Hmm I see what you mean, the first design is not so intuitive.

This forces us to go down this very odd code pattern, I am not happy 👎

Reply:
i think one had to expect such things if we were about to refactor such and unknown big beast as it was

now it is just a small beast, we know it almost all and we know how to make it even better then ...